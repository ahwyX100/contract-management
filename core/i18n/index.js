import Vue from 'vue'
import VueI18n from 'vue-i18n'

Vue.use(VueI18n)

const loadedLanguages = ['en']

const i18n = new VueI18n({
  locale: 'en',
  fallbackLocale: 'en',
  messages: {
  	'en': require('./source/en.js')
  }
})

function setI18nLanguage (lang) {
  i18n.locale = lang
  return lang
}

export function loadLanguageAsync (lang) {
  if (i18n.locale !== lang) {
    if (!loadedLanguages.includes(lang)) {
      return import(/* webpackChunkName: "lang-[request]" */ `./source/${lang}.js`).then(msgs => {
        i18n.setLocaleMessage(lang, msgs)
        loadedLanguages.push(lang)
        return setI18nLanguage(lang)
      })
    } 
    return Promise.resolve(setI18nLanguage(lang))
  }
  return Promise.resolve(lang)
}

export default i18n
