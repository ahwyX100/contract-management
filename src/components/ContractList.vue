<template>
  <div>
    <snack-bar></snack-bar>
    <v-toolbar flat color="white">
      <v-toolbar-title>{{ $t('my_contract_list') }}</v-toolbar-title>
      <v-divider
        class="mx-2"
        inset
        vertical
      ></v-divider>
      <v-spacer></v-spacer>
      <v-btn color="info" class="mb-2" @click="openModal">{{ $t('new_contract') }}</v-btn>
    </v-toolbar>
    <v-data-table
      :headers="headers"
      :items="partialContracts"
      :pagination.sync="pagination"
      :loading="loading"
      :total-items="totalCount"
      item-key="updatedTime"
      class="elevation-1"
    >
      <template slot="items" slot-scope="props">
        <td class="text-xs-left">{{ props.item.name }}</td>
        <td class="text-xs-left">{{ props.item.amountInUsd }}</td>
        <td class="text-xs-left">{{ props.item.currency }}</td>
        <td class="text-xs-left">{{ props.item.date }}</td>
        <td class="text-xs-left layout">
          <v-icon
            small
            class="mr-2"
            style="display: flex !important"
            @click="editItem(props.item)"
          >
            edit
          </v-icon>
          <v-icon
            small
            style="display: flex !important"
            @click="deleteItem(props.item)"
          >
            delete
          </v-icon>
        </td>
      </template>
      <template slot="no-data">
        <v-btn color="primary" @click="fetchData">{{ $t('Reload') }}</v-btn>
      </template>
    </v-data-table>
    <base-modal :options="modalOptions">
      <new-contract
        slot="content"
        :editedIndex="editedIndex"
        :editedItem="editedItem"
        @close="closeModal"></new-contract>
    </base-modal>
  </div>
</template>

<script>
import { createNamespacedHelpers } from 'vuex'
import NewContract from './forms/NewContract'
const { mapState, mapMutations, mapActions } = createNamespacedHelpers('contract')

const defaultPagination = {
  page: 1,
  rowsPerPage: 25
}

export default {
  name: 'ContractList',
  components: {
    NewContract
  },
  data: () => ({
    loading: true,
    pagination: Object.assign({}, defaultPagination),
    headers: [
      {
        text: 'Name',
        align: 'left',
        sortable: false,
        value: 'name'
      },
      { text: 'Amount in USD($)', value: 'amountInUsd' },
      { text: 'Currency', value: 'currency' },
      { text: 'Deal Time', value: 'date' },
      { text: 'Actions', value: 'name', sortable: false }
    ],
    modalOptions: {
      open: false,
      maxWidth: '600px'
    },
    editedIndex: -1,
    editedItem: {}
  }),

  computed: {
    ...mapState(['partialContracts', 'totalCount'])
  },

  watch: {
    pagination: {
      handler () {
        this.setPagination(this.pagination)
        this.fetchData()
      }
    },
    deep: true
  },

  created () {
    this.fetchData()
  },

  methods: {
    ...mapActions(['getContractList', 'deleteContract']),
    ...mapMutations(['setPagination']),
    fetchData () {
      this.loading = true
      this.getContractList({
        startPage: this.pagination.page,
        pageSize: this.pagination.rowsPerPage
      })
        .then(() => {
          this.loading = false
        })
    },

    editItem (item) {
      this.editedIndex = this.partialContracts.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.openModal()
    },

    deleteItem (item) {
      const targetIndex = this.partialContracts.indexOf(item)
      confirm(this.$t('confirm_delete_item')) && this.deleteContract({ id: item.id, targetIndex })
    },

    openModal () {
      this.modalOptions.open = true
    },

    closeModal () {
      this.modalOptions.open = false
      this.editedIndex = -1
    }
  }
}
</script>

<style scoped>

</style>
