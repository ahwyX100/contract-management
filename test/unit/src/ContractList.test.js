/*
 * import test utils and dependency packages
 */
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import i18n from '@nomisma/i18n'

/*
 * import test target component
 */
import ContractList from '@/components/ContractList.vue'

/*
 * mock sub component stub if child component exists
 * add plugins, directives and other dependencies locally
 */
const localVue = createLocalVue()
localVue.use(Vuex)

/*
 * main test case
 */
describe('my contract list view', () => {
  let wrapper = null

  const actions = {
    getContractList: jest.fn(),
    deleteContract: jest.fn()
  }
  const store = new Vuex.Store({
    modules: {
      contract: {
        namespaced: true,
        state: {
          contracts: [],
          partialContracts: [],
        },
        actions
      }	
    }
  })

  afterEach(() => {
    if (wrapper) {
      wrapper.destroy()
      wrapper = null
    }
  })

  function createWrapper(props, mounted) {
    return wrapper = shallowMount(ContractList, {
      propsData: props,
      localVue,
      store,
      i18n,
      sync: false,
      attachToDocument: mounted
    })
  }

  test('it renders successfully', () => {
    wrapper = createWrapper({}, true)
    expect(wrapper.element).toMatchSnapshot()
  })

  test('should fetch first data after component created', () => {
    const spy = jest.spyOn(ContractList.methods, 'fetchData')
    wrapper = createWrapper({}, true)
    expect(spy).toHaveBeenCalled()
    expect(actions.getContractList).toHaveBeenCalled()
    spy.mockRestore()
  })
})