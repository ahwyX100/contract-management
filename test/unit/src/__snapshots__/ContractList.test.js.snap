// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my contract list view it renders successfully 1`] = `
<div>
  <snack-bar-stub />
   
  <v-toolbar-stub
    color="white"
    flat="true"
    scrollthreshold="300"
  >
    <v-toolbar-title-stub>
      My Contract List
    </v-toolbar-title-stub>
     
    <v-divider-stub
      class="mx-2"
      inset="true"
      vertical="true"
    />
     
    <v-spacer-stub />
     
    <v-btn-stub
      activeclass="v-btn--active"
      class="mb-2"
      color="info"
      tag="button"
      type="button"
    >
      New Contract
    </v-btn-stub>
  </v-toolbar-stub>
   
  <v-data-table-stub
    class="elevation-1"
    customfilter="function _default(items, search, filter, headers) {
                search = search.toString().toLowerCase();
                if (search.trim() === '') return items;
                var props = headers.map(function (h) {
                    return h.value;
                });
                return items.filter(function (item) {
                    return props.some(function (prop) {
                        return filter(Object(_util_helpers__WEBPACK_IMPORTED_MODULE_7__[\\"getObjectValueByPath\\"])(item, prop, item[prop]), search);
                    });
                });
            }"
    headers="[object Object],[object Object],[object Object],[object Object],[object Object]"
    headertext="text"
    item-key="updatedTime"
    items=""
    loading="true"
    pagination="[object Object]"
    rowsperpagetext="$vuetify.dataTable.rowsPerPageText"
  >
    <template>
      <v-btn-stub
        activeclass="v-btn--active"
        color="primary"
        tag="button"
        type="button"
      >
        Reload
      </v-btn-stub>
    </template>
  </v-data-table-stub>
   
  <base-modal-stub
    options="[object Object]"
  >
    <new-contract-stub
      editedindex="-1"
      editeditem="[object Object]"
    />
  </base-modal-stub>
</div>
`;
